### PR TITLE
towncrier, gitlint: Workflow automation

### PR DESCRIPTION
**Pull request by mitjat**
This PR adds optional git hooks and scripts that make it easier for code contributors to live with `gitlint` and `markdownlint`.

### CODE REVIEW

File: .gitlint [hunk 1]
```diff
@@ -10,6 +10,9 @@ ignore-fixup-commits=false
 ignore-squash-commits=false
 ignore=body-is-missing
 contrib=contrib-disallow-cleanup-commits
+# Upcoming default; make it explicit here to avoid warnings on every gitlint run. 
+# More info: https://jorisroovers.com/gitlint/latest/configuration/general_options/#regex-style-search
+regex-style-search=true
```
Comment (mitjat) [inline .gitlint:15]: @buberdds we might want this in other repos too. The latest gitlint (0.19.1) prints an obnoxious two-line warning every time it runs, unless this config is present.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -48,8 +47,8 @@ about!
 
 If you believe a bug report presents a security risk, please follow
 [responsible disclosure](https://en.wikipedia.org/wiki/Responsible_disclosure)
-and report it directly to security@oasisprotocol.org instead of filing a public
-issue or posting it to a public forum. We will get back to you promptly.
+and report it directly to <security@oasisprotocol.org> instead of filing a
```
Comment (mitjat) [inline CONTRIBUTING.md:50]: `<>` imposed by markdownlint

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -96,11 +95,6 @@ Building and testing are documented in our
 
     _Note that you are allowed to force push into your development branches._
 
-- **Use draft pull requests for work-in-progress:**
-
-  - The draft state signals that the code is not ready for review, but still
-    gives a nice URL to track the ongoing work.
-
```
Comment (mitjat) [inline CONTRIBUTING.md:103]: This is an obvious feature of github, being used as intended. Trying to keep this document not overwhelming.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -158,23 +152,49 @@ Building and testing are documented in our
...
 A quick summary:
 
 - Separate subject from body with a blank line.
-- Limit the subject line to 80 characters.
-- Prefix the subject line with one of:
-  - "breaks:" if commit implements a non-backward compatible breaking change
-  - "fix:" if commit implements a bugfix
-  - "feat:" if commit implements a new feature
-  - "deps:" if commit updates a dependency
-  - "other:" if commit doesn't fall in any of the above categories
```
Comment (mitjat) [inline CONTRIBUTING.md:172]: We never did this. Also, it now vaguely applies to changelog fragments, not commit msgs.

File: scripts/add_changelog [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+def run_or_die(cmd: str) -> None:
+  "Runs `cmd` and dies if it fails."
+  exit_code = os.system(cmd)
+  if exit_code != 0:
+    sys.exit(exit_code)
+
+def find_pr(branch: str) -> Optional[str]:
+  "Returns the PR number for the given branch, or None if no PR exists."
+  pr, error, exit_code = popen_full(['gh', 'pr', 'view', '--json', 'number', '--jq', '.number', branch])
```
Comment (Andrew7234) [inline scripts/add_changelog:24]: Maybe we could add something in the docstring about `gh` being a prerequisite? I assume it's just `brew install gh`?

File: scripts/git-hooks/prepare-commit-msg [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  # Suppress the output of prettier, as it's overly verbose. Show only on errors.
+  prettier --no-config --no-editorconfig --prose-wrap always --print-width "$WIDTH" --write "$f" >/tmp/prettier.out \
+    || { cat /tmp/prettier.out; exit 1; }
+}
+
+# If gitlint is already happy, do nothing
+gitlint <"$COMMIT_MSG_FILE" >/tmp/gitlint.out 2>&1 && exit 0
+
+# If this is a fixup commit, do nothing
+<"$1" head -n1 | grep -q 'fixup!' && exit 0
```
Comment (Andrew7234) [inline scripts/git-hooks/prepare-commit-msg:28]: Does fixup mean anything special? Or is it similar in essence to the `nit` commits

**Review by Andrew7234 — APPROVED** [review]
Thank you for putting this together! Excited to try it out

File: scripts/add_changelog (hunk 5 above)
Comment (mitjat) [reply scripts/add_changelog:24]: I added a more informative error message (a few lines down) and I added a whole big new commit that refreshes top-level README.md. We had a section on development tools/dependencies there, and I expanded it now. Not just with `gh`; there are several other tools we're using too.

File: scripts/git-hooks/prepare-commit-msg (hunk 6 above)
Comment (mitjat) [reply scripts/git-hooks/prepare-commit-msg:28]: It's a feature of git. Fixup commits are commits with a special syntax that git recognizes and helps you use them to amend older commits. It's a very useful tool for keeping clean branches with meaningful commits (hint hint :)). This is a good intro, I have a very similar setup myself and use it many times a day; would recommend: https://jordanelver.co.uk/blog/2020/06/04/fixing-commits-with-git-commit-fixup-and-git-rebase-autosquash/
But there's a lot on the internet about fixup commits if the above is too terse.

**Comment by mitjat** [comment]
This will probably need some more tweaks in the future, but I've managed 2 PRs with it so I deem it good enough to start testing and iterating on.

File: scripts/add_changelog (hunk 5 above)
Comment (Andrew7234) [reply scripts/add_changelog]: thanks!

**Pull request merged**